homeassistant/components/sensor/__init__.py

```diff
@@ -447,3 +448,45 @@ def __repr__(self) -> str:
             return f"<Entity {self.name}>"
 
         return super().__repr__()
+
+
+@dataclass
+class SensorExtraStoredData(ExtraStoredData):
+    """Object to hold extra stored data."""
+
+    native_value: StateType | date | datetime
+    native_unit_of_measurement: str | None
+
+    def as_dict(self) -> dict[str, Any]:
+        """Return a dict representation of the sensor data."""
+        return {
+            "native_value": self.native_value,
+            "native_unit_of_measurement": self.native_unit_of_measurement,
+        }
+
+    @classmethod
+    def from_dict(cls, restored: dict[str, Any]) -> SensorExtraStoredData:
+        """Initialize a stored sensor state from a dict."""
+        return cls(restored["native_value"], restored["native_unit_of_measurement"])
+
+
+class RestoreSensor(SensorEntity, RestoreEntity):
+    """Mixin class for restoring previous sensor state."""
+
+    @property
+    def extra_restore_state_data(self) -> Mapping[str, SensorExtraStoredData] | None:
+        """Return sensor specific state data to be restored."""
+        return {
+            "sensor": SensorExtraStoredData(
+                self.native_value, self.native_unit_of_measurement
+            )
```

Comment:
The idea here is to allow integrations deriving entities from `RestoreSensor` to add additional keys.
Maybe that's unnecessary though, and such an entity should just return an instance of a subclassed `SensorExtraStoredData` instead.

Reply:
If the platform still needs to return an instance of `SensorExtraStoredData` as value in the mapping, adding more keys without subclassing `SensorExtraStoredData` is probably not useful. Or do you have an example already?

Reply:
We should change the typing of `RestoreSensor.extra_restore_state_data` function to be:
```py
def extra_restore_state_data(self) -> Mapping[str, ExtraStoredData] | None:
```

It's error prone though because we would rely on all implementations to cooperate and create a union dict.
Maybe better to simply retype `RestoreEntity.extra_restore_state_data` as
```py
def extra_restore_state_data(self) -> ExtraStoredData | None:
```

Reply:
The latter sounds good. 👍

Reply:
OK, fixed in 97d1156

I think we need some safeguarding when attempting to deserialize to `SensorExtraStoredData`.
Maybe we should include class name + version in the saved data and check for that when restoring?

Reply:
Yeah. Maybe just check if the parameter keys are present and if if not return None?